### DISCUSSION TITLE
How do you assert emitSelf?

### DISCUSSION THREAD

**Discussion by trip-somers** [livewire/livewire#5455] [1. Help]
I currently have a component with the following line in a Create component:

```php
$this->emitSelf('continue', '/organizations/' . $model->organization_id);
```

In CreateTest, I'm trying the following:

```php
$this->livewire(Create::class)
    // ... set data ...
    ->call('store')
    ->assertEmittedTo(Create::class, 'continue', '/organizations/63a2405ef6ea169814046497');
```

This fails to recognize the emitted event with the following message: `Failed asserting that an event [continue] was fired with parameters: ["\/organizations\/63a2405ef6ea169814046497"]`.

There is no `->assertEmittedSelf()` method, and I feel like a generic `->assertEmitted()` isn't specific enough.

UPDATE #1: Even after switching to `->assertEmitted()`, it is still failing with the same message. Is it rejecting it because of the escape characters???

UPDATE #2: I was only able to get this to pass with a generic `->assertEmitted(Create::class, 'continue')` assertion, which feels not close to specific enough.

**Comment by joshhanley** [comment]
@trip-somers I think the only way is using `assertEmitted` as you have found.

I've never found a reason to use `emitSelf` though, as I just normally just run whatever code I need inside the component where you would `emitSelf`, as `emitSelf` triggers a second network request which seems unnecessary.

Are you able to share some details on what you are using `emitSelf` for?

**Reply by trip-somers** [reply]
It's related to how we're handling this issue: https://github.com/livewire/livewire/discussions/5424

At one point, we set up our `store()` method to check for a not-null `result` with which we were initiating the redirect by triggering a request like this: `$wire.continue(result)`. For this to work, we were returning a target URL for a redirect from our Component's store() method.

That worked for what we needed, but it felt weird to have the Alpine method interpret the result and then make a follow-up Livewire request. I figured I could accomplish the same thing by having the component emit the `continue` event to itself:

```php
    protected function store()
    {
        $this->model->save();

        $this->emitSelf(
            'continue',
            route('models.show', [
                'model'    => $this->model->id,
            ])
        );
    }
```

Because the event (`continue`) is so generic, we want to use emitSelf() to prevent any other components from responding to it now or in the future as the app evolves.

**Reply by trip-somers** [reply]
The reason we switched from the Alpine/$wire relay was because we could not assert that the redirect was happening. With event emission, we could assert that one method emitted it and another method listened for it.

**Reply by titantwentyone** [reply]
Just wanted to follow up on this as I've been looking for the same thing. My implementation is a cart where each `CartItem` livewire component has a method `remove`. I can confirm that a `removeCartItem` event is triggered but I can't test that it has only been sent to itself.

> I've never found a reason to use emitSelf though, as I just normally just run whatever code I need inside the component where you would emitSelf, as emitSelf triggers a second network request which seems unnecessary.

In my use case, I'm using a notification system provided by a third party (Filament) which asks the customer "Do you want to remove the item?". If affirmative, then an event is triggered so I can't just roll it up into a method on the component.

We do have `assertEmittedTo` so we are able to confirm that an event was handed off to any other component. Looking through the source just now, I see we have an `assertEmittedUp` also. It feels like `assertEmittedSelf` would plug the one final area where we can accurately test to what and with what an event is transmitted to.

Happy to have a bash at a PR?

**Reply by joshhanley** [reply]
Yeah makes sense. Feel free to submit a PR 🙂 just a heads up, we are limiting changes in V2 as much as we can while we work on V3, so it may not necessarily be accepted at the moment. But I'd say submit it anyway so Caleb can have a look and decide if he wants it in now or not.

**Reply by titantwentyone** [reply]
PR #5610
